cmd/secrets/secret_helper.go

```diff
@@ -53,105 +82,102 @@ type secretsRequest struct {
+		return nil, fmt.Errorf("incompatible protocol version %q", request.Version)
 	}
 
 	if len(request.Secrets) == 0 {
-		return errors.New("no secrets listed in input")
+		return nil, errors.New("no secrets listed in input")
 	}
 
-	response := map[string]s.Secret{}
-	for _, name := range request.Secrets {
-		response[name] = readSecret(filepath.Join(dir, name))
-	}
+	return request.Secrets, nil
+}
 
-	out, err := json.Marshal(response)
+func writeFetchedSecrets(w io.Writer, fetchedSecrets map[string]s.Secret) error {
+	out, err := json.Marshal(fetchedSecrets)
 	if err != nil {
 		return err
 	}
+
 	_, err = w.Write(out)
 	return err
 }
 
-func splitVersion(ver string) []string {
-	return strings.SplitN(ver, ".", 2)
-}
+func readSecretsFromFile(secrets []string, dir string) map[string]s.Secret {
+	res := make(map[string]s.Secret)
 
-func readSecret(path string) s.Secret {
-	value, err := readSecretFile(path)
-	var errMsg string
-	if err != nil {
-		errMsg = err.Error()
+	secretsProvider := providers.NewFileProvider(dir)
+	for _, secretID := range secrets {
+		res[secretID] = secretsProvider.Get(secretID)
 	}
-	return s.Secret{Value: value, ErrorMsg: errMsg}
+
+	return res
 }
 
-func readSecretFile(path string) (string, error) {
-	fi, err := os.Lstat(path)
-	if err != nil {
-		if os.IsNotExist(err) {
-			return "", errors.New("secret does not exist")
-		}
-		return "", err
-	}
+func readSecretsUsingPrefixes(secrets []string, selector *providers.Selector) map[string]s.Secret {
+	res := make(map[string]s.Secret)
 
-	if fi.Mode()&os.ModeSymlink != 0 {
-		// Ensure that the symlink is in the same dir
-		target, err := os.Readlink(path)
+	for _, secretID := range secrets {
+		prefix, id, err := parseSecretWithPrefix(secretID)
 		if err != nil {
-			return "", fmt.Errorf("failed to read symlink target: %v", err)
-		}
-
-		dir := filepath.Dir(path)
-		if !filepath.IsAbs(target) {
-			target, err = filepath.Abs(filepath.Join(dir, target))
-			if err != nil {
-				return "", fmt.Errorf("failed to resolve symlink absolute path: %v", err)
-			}
+			res[secretID] = s.Secret{Value: "", ErrorMsg: err.Error()}
+			continue
 		}
 
-		dirAbs, err := filepath.Abs(dir)
+		secretsProvider, err := selector.GetProvider(prefix)
 		if err != nil {
-			return "", fmt.Errorf("failed to resolve absolute path of directory: %v", err)
+			res[secretID] = s.Secret{Value: "", ErrorMsg: err.Error()}
+			continue
 		}
 
-		if !filepath.HasPrefix(target, dirAbs) {
-			return "", fmt.Errorf("not following symlink %q outside of %q", target, dir)
-		}
-	}
-	fi, err = os.Stat(path)
-	if err != nil {
-		return "", err
+		res[secretID] = secretsProvider.Get(id)
 	}
 
-	if fi.Size() > maxSecretFileSize {
-		return "", errors.New("secret exceeds max allowed size")
-	}
+	return res
+}
 
-	file, err := os.Open(path)
-	if err != nil {
-		return "", err
+func parseSecretWithPrefix(secretID string) (prefix providers.Prefix, id string, err error) {
```

Comment:
It would be nice if this function hand the case where the prefix is not here. And in that case default to the "legacy" provider (fromFile). That is also why I was thinking to use a separator that is not part of a path (see previous comment).

I'm asking it, because I think of the migration path. I don't think we can ask a user to switch all the secret definitions present in the AD configurations on in the files at the same time that they switch to the secret backend implementation.

Of course if we don't detect the prefix, the function can log a Warning to ask the user to update the secret configuration.

WDTY?

Reply:
I think it might make things easier for customers using a large number of secrets :+1:
I added a new commit "[secret_helper] Provide migration path from old exec" with this.